Julia: Add ex1a tic-tac-toe (#140)
exercism/research_experiment_1: exercises/julia-1-a/README.md

## Instructions

In this exercise, you're going to implement a program that determines the state of a [tic-tac-toe](https://en.wikipedia.org/wiki/Tic-tac-toe) game.
(_You may also know the game as "noughts and crosses" or "Xs and Os"._)

The games is played on a 3x3 grid.
Players take turns to place `X`s and `O`s on the grid.
The game ends when one player has won by placing three of marks in a row, column, or along a diagonal of the grid, or when the entire grid is filled up.

> Without loss of generality we will assume that `X` starts.

Reviewer: I don't know what this means. I'm also not sure it's needed for this part?

Author: It's needed because it decides which boards are invalid due to taking turns out of order. I can reduce it to

```suggestion
For the entire exercise, we will assume that `X` starts.
```

The w.l.o.g. means that we don't have to consider games where `O` starts because the solution/implementation would be nearly identical with the only exception that you'd first replace `X`->`O` and `O`->`X`.

Reviewer 2: I also found the w.l.o.g. a bit hard to read. I like the suggested change!

## Instructions

> Implement your own [tic-tac-toe](https://en.wikipedia.org/wiki/Tic-tac-toe) game.

Reviewer: I think the task is to implement a function to determine "state" of game, not create the game itself (which to me implied flow and state, etc).

Other than that, this is good! :)

Author: ```suggestion
In this exercise, you will implement a computer-controlled [tic-tac-toe](https://en.wikipedia.org/wiki/Tic-tac-toe) opponent.
In the first part, you will write a program that analyses the state of the game.
```

Is this better?

Reviewer: No :) 

We're not telling people there **is** a second part, because that then makes them code with future requirements in mind, which invalidates the research. So all we want to mention is the state of game bit :)

Author: Ah, I didn't know that. How about this?

```suggestion
In this exercise, you're going to implement a program that determines the state of a [tic-tac-toe](https://en.wikipedia.org/wiki/Tic-tac-toe) game.
```

Reviewer: Thanks :)

(_You may also know the game as "noughts and crosses" or "Xs and Os"._)

The games is played on a 3x3 grid.
Players take turns to place `X`s and `O`s on the grid.
The game ends when one player has won by placing three of marks in a row, column, or along a diagonal of the grid, or when the entire grid is filled up.

Without loss of generality we will assume that `X` starts.

> At first, you will create create a function `gamestate` that takes the game board represented a string as its sole argument and returns a Symbol describing the state of the game.

Author: ```suggestion
Create a function `gamestate` that takes the game board represented a string as its sole argument and returns a Symbol describing the state of the game.
```

The board will be given as a 3-line string containing `X`s, `O`s and ` `s:

```julia
board = "X  \n XO\nOX "
```

represents the following board:

```
   |   |   
 X |   |   
___|___|___
   |   |   
   | X | O 
___|___|___
   |   |
 O | X |  
   |   |
```

There are 3 potential game states:

- The game is **ongoing**.
- The game ended in a **draw**.
- The game ended in a **win**.

> If the given board is invalid, throw an appropriate error.

Author: Do you think this is clear enough or should I explain what kinds of boards would be invalid?

The tests check for taking turns out of order or boards that have multiple rows/cols/diags with 3 in a row.

Reviewer 2: It might be good to explicitly list the invalid states, as that means less thinking by the student and more coding (which is what we want).

Author: Done in 1414171

> If the given board is invalid, throw an appropriate error.

Author: ```suggestion
If the given board is invalid, throw an appropriate error.

If a board meets the following conditions, it is invalid:

- The players took turns out of order (remember that `X` starts).
- The game was played after it already ended, i.e. more than one row/column/diagonal contains the same symbol thrice.
```
